macros/src/pubsub.rs

```diff
@@ -37,16 +40,64 @@ impl<T> Subscriber<T> {
 pub struct Sink<T> {
 	sink: pubsub::Sink,
 	id: SubscriptionId,
+	buffered: Option<core::Params>,
 	_data: PhantomData<T>,
 }
 
 impl<T: serde::Serialize> Sink<T> {
-	pub fn send(&self, val: T) -> pubsub::SinkResult {
+	pub fn notify(&self, val: T) -> pubsub::SinkResult {
+		self.sink.notify(self.val_to_params(val))
+	}
+
+	fn val_to_params(&self, val: T) -> core::Params {
 		let id = self.id.clone().into();
 		let val = to_value(val);
-		self.sink.send(core::Params::Map(vec![
+
+		core::Params::Map(vec![
 			("subscription".to_owned(), id),
 			("result".to_owned(), val),
-		].into_iter().collect()))
+		].into_iter().collect())
+	}
+
+	fn poll(&mut self) -> futures::Poll<(), pubsub::TransportError> {
+		if let Some(item) = self.buffered.take() {
+			let result = self.sink.start_send(item)?;
+			if let futures::AsyncSink::NotReady(item) = result {
+				self.buffered = Some(item);
+			}
+		}
+
+		if self.buffered.is_some() {
+			Ok(futures::Async::NotReady)
+		} else {
+			Ok(futures::Async::Ready(()))
+		}
+	}
+}
+
+impl<T: serde::Serialize> futures::sink::Sink for Sink<T> {
+	type SinkItem = T;
+	type SinkError = pubsub::TransportError;
+
+	fn start_send(&mut self, item: Self::SinkItem) -> futures::StartSend<Self::SinkItem, Self::SinkError> {
+		if self.poll()?.is_not_ready() {
+			return Ok(futures::AsyncSink::NotReady(item));
```

Comment:
I'm pretty sure it doesn't, but putting the rationale why in a comment above (since futures have a lot of hidden complexity) would help